src/master-playlist-controller.js

```diff
@@ -1235,17 +1242,35 @@ export class MasterPlaylistController extends videojs.EventTarget {
   }
 
   /**
-   * Calculates the desired forward buffer length based on current time
+   * Calculates or sets the desired forward buffer length based on current time
    *
-   * @return {Number} Desired forward buffer length in seconds
+   * @param {Number=} Desired forward buffer length in seconds
+   * @return {Number} Current forward buffer target length
    */
-  goalBufferLength() {
-    const currentTime = this.tech_.currentTime();
-    const initial = Config.GOAL_BUFFER_LENGTH;
-    const rate = Config.GOAL_BUFFER_LENGTH_RATE;
-    const max = Math.max(initial, Config.MAX_GOAL_BUFFER_LENGTH);
+  goalBufferLength(value) {
+    if (value === undefined) {
+      const currentTime = this.tech_.currentTime();
+      const initial = this.goalBufferLength_;
+      const rate = Config.GOAL_BUFFER_LENGTH_RATE;
+      const max = Math.max(initial, this.maxGoalBufferLength_);
+
+      return Math.min(initial + currentTime * rate, max);
+    }
+    this.goalBufferLength_ = Math.min(this.goalBufferLength_, value);
+    this.maxGoalBufferLength_ = Math.min(this.maxGoalBufferLength_, value);
```

Comment:
> shouldn't this be the max?

No. The `Math.min` guarantees that if we have to shorten the buffer due to a `QuotaExceededError`, it never enlarges again. It may not be _strictly_ necessary, but I opted for safety.

If it was `Math.max`, the `maxGoalBufferLength` would never change.

> Further, why is the maxGoalBufferLength changing at all?

The buffer lengths are adjusted so that we don't keep calling the event loop. If we don't adjust the buffers, then `checkBuffer_` gets called way more frequently, and just keeps exiting without downloading because it would see that the next segment would bust the buffer (in terms of bytes). Adjusting the buffer sizes means we can actually meet the desired buffer length, and cool off the `checkBuffer_` calls.

Reply:
Just to hopefully clarify, when using this function as a setter, you're saying that's the largest you ever want `goalBufferLength_` to be. It could be _smaller_, but it will never be _larger_.

It will be smaller if `initial + currentTime * rate` is less than the max. The buffer length and the water line scale up over time from start (`t=0`) to some time later (`t=n`). `goalBufferLength` is the value at `t=0` and `maxGoalBufferLength` is the value at `t=n`.

Reply:
I think you should consult #1147 and #1134 since I am merely integrating with pre-existing functionality. My understanding is that the way things are currently working, is the desired behaviour -- abstracting away a growing buffer goal. The other members of the team who made that functionality are probably in a better place to answer your more in-depth questions than I am. 

> Does that mean that both the `goalBufferLength` and `maxGoalBufferLength` are exactly the same?

Yes.

> Is that ideal?

There's no room left to scale, so... :man_shrugging:  I think that's another question for the original authors.

Reply:
After discussing this out-of-band:
- `goalBufferLength()` will return a growing number for the forward buffer goal until a QuotaExceededError is seen, at which point we'll dynamically change the forward buffer goal and max forward buffer goal to take into account the browser's max source buffer allowance
- we can't rename `goalBufferLength` for backwards compatibility